incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -1690,80 +1758,93 @@ else if (this.authField.equals(MqttConnectProperty.PASSWORD))
-                }
-                else
+
+                if (!session)
                 {
-                    doEncodeConnack(traceId, authorization, reasonCode, assignedClientId, false, null);
+                    doEncodeConnack(traceId, authorization, reasonCode, assignedClientId,
+                        false, null);
                     connected = true;
                 }
 
-                doCancelConnectTimeout();
+                final int flags = connectFlags;
+                final boolean willFlagSet = isSetWillFlag(flags);
+                final int willFlags = decodeWillFlags(flags);
+                final int willQos = decodeWillQos(flags);
 
-                decoder = decodePacketType;
-            }
+                if (session && willFlagSet)
+                {
+                    final MqttDataExFW.Builder sessionDataExBuilder =
+                        mqttSessionDataExRW.wrap(sessionExtBuffer, 0, sessionExtBuffer.capacity())
+                            .typeId(mqttTypeId)
+                            .session(s -> s.kind(k -> k.set(MqttSessionDataKind.WILL)));
 
-            if (reasonCode == BAD_USER_NAME_OR_PASSWORD)
-            {
-                doCancelConnectTimeout();
-                doNetworkEnd(traceId, authorization);
-                decoder = decodeIgnoreAll;
-                progress = connect.limit();
+                    final MqttMessageFW.Builder willMessageBuilder =
+                        mqttMessageFW.wrap(willMessageBuffer, 0, willMessageBuffer.capacity())
+                            .topic(payload.willTopic)
+                            .delay(payload.willDelay)
+                            .qos(willQos)
+                            .flags(willFlags)
+                            .expiryInterval(payload.expiryInterval)
+                            .contentType(payload.contentType)
+                            .format(f -> f.set(payload.payloadFormat))
+                            .responseTopic(payload.responseTopic)
+                            .correlation(c -> c.bytes(payload.correlationData));
+
+                    final Array32FW<MqttUserPropertyFW> userProperties = willUserPropertiesRW.build();
+                    userProperties.forEach(
+                        c -> willMessageBuilder.propertiesItem(p -> p.key(c.key()).value(c.value())));
+                    willMessageBuilder.payload(p -> p.bytes(payload.willPayload.bytes()));
+
+                    final MqttMessageFW will = willMessageBuilder.build();
+                    final int willPayloadSize = willMessageBuilder.sizeof();
+
+                    if (!sessionStream.hasSessionWindow(willPayloadSize))
+                    {
+                        break decode;
+                    }
+                    sessionStream.doSessionData(traceId, willPayloadSize, sessionDataExBuilder.build(), will);
+                }
+                decodableRemainingBytes -= connectPayloadLimit - progress;
```

Comment:
Please move this manipulation of `decodableRemainingBytes` to `decodeConnectPayload` for better separation of concerns, much like no manipulating the `decoder` here.